internal/sops/awskms/keysource.go

```diff
@@ -0,0 +1,271 @@
+/*
+Copyright 2022 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awskms
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"encoding/base64"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/config"
+	"github.com/aws/aws-sdk-go-v2/credentials"
+	"github.com/aws/aws-sdk-go-v2/service/kms"
+	"github.com/aws/aws-sdk-go-v2/service/sts"
+	"sigs.k8s.io/yaml"
+)
+
+const (
+	// valid ARN example: arn:aws:kms:us-west-2:107501996527:key/612d5f0p-p1l3-45e6-aca6-a5b005693a48
+	arnRegex = `^arn:aws[\w-]*:kms:(.+):[0-9]+:(key|alias)/.+$`
+	// valid STS session examples: john_s, sops@42WQm042
+	stsSessionRegex = "[^a-zA-Z0-9=,.@-_]+"
+	// kmsTTL is the duration after which a MasterKey requires rotation.
+	kmsTTL = time.Hour * 24 * 30 * 6
+)
+
+// MasterKey is an AWS KMS key used to encrypt and decrypt sops' data key.
+// Adapted from: https://github.com/mozilla/sops/blob/v3.7.2/kms/keysource.go#L39
+// Modified to accept custom static credentials as opposed to using env vars by default
+// and use aws-sdk-go-v2 instead of aws-sdk-go being used in upstream.
+type MasterKey struct {
+	Arn               string
+	Role              string
+	EncryptedKey      string
+	CreationDate      time.Time
+	EncryptionContext map[string]string
+
+	credentialsProvider aws.CredentialsProvider
+
+	// epResolver IS ONLY MEANT TO BE USED FOR TESTS.
+	// it can be used to override the endpoint that the AWS client resolves to
+	// by default. it's hacky but there is no other choice, since you can't
+	// specify the endpoint as an env var like you can do with an access key.
+	epResolver aws.EndpointResolver
```

Comment:
```suggestion
	// epResolver can be used to override the endpoint that the AWS client
	// resolves to by default. This is mostly used for testing purposes as it can
	// not be injected using e.g. an environment variable. The field is not
	// publicly exposed, nor configurable.
	epResolver aws.EndpointResolver
```